### PR TITLE
feat(Facade): Add support for Uint8Array and ArrayBuffer

### PR DESCRIPTION
**Pull request by jteplitz**

### CODE REVIEW

File: lib/facade_converter.ts [hunk 1]
```diff
@@ -92,19 +92,33 @@ export class FacadeConverter extends base.TranspilerBase {
...
       }
-      return n.getChildren(sourceFile).some(hasXmlHttpRequest);
-    };
+    }
+  }
 
-    if (hasXmlHttpRequest(sourceFile)) {
-      this.emit('import "dart:html";');
+  private _hasType(n: ts.Node, typeName: String, sourceFile?: ts.SourceFile): boolean {
+    if (n.kind === ts.SyntaxKind.TypeReference) {
```
Comment (mprobst) [inline lib/facade_converter.ts]: This walks the entire AST for a source file n times for each of the - currently three - types that need special casing. Can we do this in just one walk, e.g. by keeping track of what extra imports we emitted already, and then emitting directly during the search?

Comment (jteplitz602) [inline lib/facade_converter.ts]: Good idea. Just updated emitImports to do just that.

File: lib/facade_converter.ts [hunk 2]
```diff
@@ -92,19 +95,34 @@ export class FacadeConverter extends base.TranspilerBase {
...
+    var libraries = <ts.Map<string>>{
+      "XMLHttpRequest": "dart:html",
+      "Uint8Array": "dart:typed_arrays",
+      "ArrayBuffer": "dart:typed_arrays"
+    };
+    var emitted: Set = {};
+    this._emitImports(sourceFile, libraries, emitted);
+  }
+
+  private _emitImports(n: ts.Node, libraries: ts.Map<string>, emitted: Set,
```
Comment (alexeagle) [inline lib/facade_converter.ts]: I don't think we typically use underscore naming in ts2dart (it's pure typescript so the private keyword is sufficient) - can you check that and remove if this is the first underscore prefix being introduced?

File: lib/facade_converter.ts [hunk 3]
```diff
@@ -92,19 +95,34 @@ export class FacadeConverter extends base.TranspilerBase {
...
-        var type = base.ident((<ts.TypeReferenceNode>n).typeName);
-        if (type === 'XMLHttpRequest') {
-          return true;
+    var libraries = <ts.Map<string>>{
+      "XMLHttpRequest": "dart:html",
+      "Uint8Array": "dart:typed_arrays",
+      "ArrayBuffer": "dart:typed_arrays"
+    };
+    var emitted: Set = {};
+    this.emitImports(sourceFile, libraries, emitted);
```
Comment (mprobst) [inline lib/facade_converter.ts]: I find not passing `sourceFile` in this call and then doing the cast below quite confusing. Could you make the parameter non-optional and just pass `sourceFile` twice here?

File: lib/facade_converter.ts [hunk 4]
```diff
@@ -92,19 +95,34 @@ export class FacadeConverter extends base.TranspilerBase {
...
+    }
 
-    if (hasXmlHttpRequest(sourceFile)) {
-      this.emit('import "dart:html";');
+    if (!sourceFile) {
+      sourceFile = <ts.SourceFile>n;
     }
+
+    n.getChildren(sourceFile)
+        .forEach((n: ts.Node) => { this.emitImports(n, libraries, emitted, sourceFile); });
```
Comment (mprobst) [inline lib/facade_converter.ts]: I think you can leave off the curly braces here.

File: test/facade_converter_test.ts [hunk 5]
```diff
@@ -73,6 +73,10 @@ describe('type based translation', () => {
       expectWithTypes('var n: Node;').to.equal(' dynamic n ;');
       expectWithTypes('var _xhr: XMLHttpRequest;')
           .to.equal(' import "dart:html"; HttpRequest _xhr ;');
+      expectWithTypes('var _intArray: Uint8Array;')
```
Comment (mprobst) [inline test/facade_converter_test.ts]: could you remove the leading underscores from the variables? Please also remove them on the line above.

**Comment by mprobst** [comment]
@jteplitz602 I noticed this was lying around, going stale. Please remember to comment on the pull request so that your reviewers get notified (something like *"PTAL!"* is enough).

**Comment by jteplitz602** [comment]
Got it. I implemented the changes you suggested. Can you take another look? Also, Travis hasn't kicked in since my latest commit for some reason.

**Comment by mprobst** [comment]
LGTM, please merge.

**Comment by jteplitz602** [comment]
Great. I don't have write access to this repo. Can you merge?

**Comment by alexeagle** [comment]
merged in 124f126

**Pull request closed without merging**